Modules/_pickle.c

```diff
@@ -1311,6 +1339,74 @@ _Unpickler_ReadImpl(UnpicklerObject *self, char **s, Py_ssize_t n)
     return n;
 }
 
+/* Read `n` bytes from the unpickler's data source, storing the result in `buf`.
+ *
+ * This should only be used for non-small data reads where potentially
+ * avoiding a copy is beneficial.  This method does not try to prefetch
+ * more data into the input buffer.
+ *
+ * _Unpickler_Read() is recommended in most cases.
+ */
+static Py_ssize_t
+_Unpickler_ReadInto(UnpicklerObject *self, char *buf, Py_ssize_t n)
+{
+    assert(n != READ_WHOLE_LINE);
+
+    /* XXX add tests from ogrisel's use case */
+
+    /* Read from available buffer data, if any */
+    Py_ssize_t in_buffer = self->input_len - self->next_read_idx;
+    if (in_buffer > 0) {
+        Py_ssize_t to_read = Py_MIN(in_buffer, n);
+        memcpy(buf, self->input_buffer + self->next_read_idx, to_read);
+        self->next_read_idx += to_read;
+        buf += to_read;
+        n -= to_read;
+        if (n == 0) {
+            /* Entire read was satisfied from buffer */
+            return n;
+        }
+    }
+
+    /* Read from file */
+    if (_Unpickler_SkipConsumed(self) < 0)
+        return -1;
+
+    Py_ssize_t read_size;
+
+    if (self->readinto == NULL) {
+        /* Call read() and memcpy */
+        /* XXX do we want this fallback? pickle.py doesn't have it */
```

Comment:
@ogrisel Do you think we need to support "file-like" objects without a readinto() method?

Reply:
I don't have a strong opinion. Are there any such file objects in the standard library? The file-like objects that acts as wrappers for (de)compression algorithm such as `GzipFile` seem to support `readinto`.

Reply:
There doesn't seem to be any. I'll remove this snippet.

Reply:
Actually, the doc mentions that API contract for `file` is only to have `read` and `readline`, so I'm not sure it's ok to remove the fallback...

Reply:
Was the fallback ever added back? Looks like this came up as a real issue on https://bugs.python.org/issue39681

Reply:
Thanks for the ping. I'll reply on the issue.